testy/runtime.py

```diff
@@ -1,25 +1,103 @@
-import threading
+import multiprocessing
 import time
-import studentCode
+import sys
+import traceback
+
+import stateManager
+import studentAPI
+
+from runtimeUtil import *
+
+
+# TODO:
+# 0. Set up testing code for the following features.
+# DONE 1. Have student code go through api to modify state.
+# DONE 2. Imposing timeouts on student code (infinite loop, try-catch)
+# 3. Figure out how to kill student thread.
+# 4. Integrate with Bob's socket code: spin up a communication process
+# 5. stateManager throw badThing on processNameNotFound
+# 6. refactor process startup code: higher order function
+# 7. Writeup how all this works
+# 8. Investigate making BadThing extend exception
+
+allProcesses = {}
+badThings = multiprocessing.Condition()
+globalBadThing = "unititialized globalBadThing"
 
-STUDENT_THREAD_NAME = "student_thread"
 
 def runtime():
-  state = [0]
-  runStudentCode(state)
-  while True:
-    time.sleep(1)
-
-def runStudentCode(state):
-  studentThread = threading.Thread(target=runStudentCodeHelper, name=STUDENT_THREAD_NAME, args=(state,))
-  studentThread.daemon = True
-  studentThread.start()
-
-def runStudentCodeHelper(state):
-  nextCall = time.time()
-  while True:
-    studentCode.main(state)
-    nextCall += 1
-    time.sleep(nextCall - time.time())
+  badThingsQueue = multiprocessing.Queue()
+  stateQueue = multiprocessing.Queue()
+  spawnProcess = processFactory(badThingsQueue, stateQueue)
+  restartCount = 0
```

Comment:
There should probably be a restart counter reset at some point, for example if student code runs successfully. That way the code gives up only if it keeps restarting due to the same error, and not if it's just left running for a long time.